File: algorithms/java/src/test/java/leetcode/LC046Permutations.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.*;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/permutations/
 * #medium
 */
public final class LC046Permutations {
    public final class Solution {
        public void permute(final List<Integer> prefix, final Set<Integer> remaining, final List<List<Integer>> permutations) {
            if (remaining.size() == 0) {
                permutations.add(new ArrayList(prefix));
            } else {
                int size = prefix.size();
                for (int value : remaining) {
                    prefix.add(value);
                    Set<Integer> reduced = new HashSet(remaining);
                    reduced.remove(value);
                    permute(prefix, reduced, permutations);
                    prefix.remove(size);
                }
            }
        }

        public List<List<Integer>> permute(final int[] nums) {
            List<List<Integer>> permutations = new LinkedList<>();
            Set<Integer> remaining = new HashSet();
            for (int value : nums) {
                remaining.add(value);
            }
            permute(new LinkedList(), remaining, permutations);
            return permutations;
        }
    }

    private class IntegerListComparator implements Comparator<List<Integer>>, Serializable {
        @Override
        public int compare(final List<Integer> l1, final List<Integer> l2) {
            if (l1.size() < l2.size()) return -1;
            if (l1.size() > l2.size()) return 1;
            for (int i = 0; i < l1.size(); i++) {
                if (l1.get(i) < l2.get(i)) return -1;
                if (l1.get(i) > l2.get(i)) return 1;
            }
            return 0;
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] nums = {1, 2, 3};
        int[][] expected = {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}};
        List<List<Integer>> result = new Solution().permute(nums);
        Collections.sort(result, new IntegerListComparator());
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i].length, result.get(i).size());
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], result.get(i).get(j).intValue());
            }
        }
    }
}

File: algorithms/code/leetcode/lc046_permutations/lc046_permutations.py
# https://leetcode.com/problems/permutations/

import unittest
from typing import Any, Callable


class Solution:
    def __permute(
        self,
        prefix: list[int],
        remaining: set[int],
        permutations: list[list[int]],
    ) -> None:
        if len(remaining) == 0:
            permutations.append(prefix.copy())
        else:
            for value in remaining:
                prefix.append(value)
                reduced = set(remaining)
                reduced.remove(value)
                self.__permute(prefix, reduced, permutations)
                prefix.pop()

    def permute(self, nums: list[int]) -> list[list[int]]:
        permutations: list[list[int]] = []
        remaining = set(nums)
        self.__permute([], remaining, permutations)
        return permutations


class TestCode(unittest.TestCase):
    @staticmethod
    def cmp_to_key(mycmp: Callable) -> Any:  # pragma: no cover
        class K:
            def __init__(self, obj: Any, *_: Any) -> None:
                self.obj = obj

            def __lt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) < 0

            def __gt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) > 0

            def __eq__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) == 0

            def __le__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) <= 0

            def __ge__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) >= 0

            def __ne__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) != 0

        return K

    @staticmethod
    def integer_list_comparator(
        l1: list[int], l2: list[int]
    ) -> int:  # pragma: no cover
        if len(l1) < len(l2):
            return -1
        if len(l1) > len(l2):
            return 1
        for l1i, l2i in zip(l1, l2, strict=True):
            if l1i < l2i:
                return -1
            if l1i > l2i:
                return 1
        return 0

    def test_example(self) -> None:
        nums = [1, 2, 3]
        expected = [
            [1, 2, 3],
            [1, 3, 2],
            [2, 1, 3],
            [2, 3, 1],
            [3, 1, 2],
            [3, 2, 1],
        ]
        result = Solution().permute(nums)
        result = sorted(
            result, key=TestCode.cmp_to_key(TestCode.integer_list_comparator)
        )
        assert len(expected) == len(result)
        for i, expected_i in enumerate(expected):
            assert len(expected_i) == len(result[i])
            for j, expected_i_j in enumerate(expected_i):
                assert expected_i_j == result[i][j]
